src/TypeSystem/Inferrer/TypeInferrer.php

```diff
@@ -0,0 +1,97 @@
+ * so it matches the expected runtime behaviour
+ *
+ * For example given this expression: `nullableString ? "nullableString is not null" : "nullableString is null"` based on the condition `nullableString`
+ * It will infer that in the truthy context nullableString is a string while in the falsy context it will infer that it is a null
+ *
+ * The structure is partially inspired by phpstan
+ * https://github.com/phpstan/phpstan-src/blob/07bb4aa2d5e39dafa78f56c5df132c763c2d1b67/src/Analyser/TypeSpecifier.php#L111
+ */
+class TypeInferrer
+{
+    public function __construct(
+        private readonly ScopeInterface $scope
+    ) {
+    }
+
+    public function inferTypesInCondition(ExpressionNode $conditionNode, TypeInferrerContext $context): InferredTypes
+    {
+        if ($conditionNode->root instanceof IdentifierNode) {
+            $type = $this->scope->lookupTypeFor($conditionNode->root->value);
+            if (!$type) {
+                return InferredTypes::empty();
+            }
+            // case `nullableString ? "nullableString is not null" : "nullableString is null"`
+            return InferredTypes::fromType($conditionNode->root->value, $context->narrowDownType($type));
+        }
+
+        if (($binaryOperationNode = $conditionNode->root) instanceof BinaryOperationNode) {
+            // cases
+            // `nullableString === null ? "nullableString is null" : "nullableString is not null"`
+            // `nullableString !== null ? "nullableString is not null" : "nullableString is null"`
+            if (count($binaryOperationNode->operands->rest) !== 1) {
+                return InferredTypes::empty();
+            }
+            $first = $binaryOperationNode->operands->first;
+            $second = $binaryOperationNode->operands->rest[0];
+
+            $comparedIdentifierValueToNull = match (true) {
+                // case `nullableString === null`
+                $first->root instanceof IdentifierNode && $second->root instanceof NullLiteralNode => $first->root->value,
+                // yodas case `null === nullableString`
+                $first->root instanceof NullLiteralNode && $second->root instanceof IdentifierNode => $second->root->value,
+                default => null
+            };
+
+            if ($comparedIdentifierValueToNull === null) {
+                return InferredTypes::empty();
+            }
+            $type = $this->scope->lookupTypeFor($comparedIdentifierValueToNull);
+            if (!$type) {
+                return InferredTypes::empty();
+            }
+
+            if ($binaryOperationNode->operator === BinaryOperator::EQUAL) {
+                return InferredTypes::fromType($comparedIdentifierValueToNull, $context->negate()->narrowDownType($type));
+            }
+            if ($binaryOperationNode->operator === BinaryOperator::NOT_EQUAL) {
+                return InferredTypes::fromType($comparedIdentifierValueToNull, $context->narrowDownType($type));
+            }
+        }
```

Comment:
These cases may be split into separate `*Narrower` classes - similarly to how the `*Resolver`-classes are organized. But I think it's okay to leave them in this class for now.

Reply:
I will wait till this class is more complex to see what other usecase we also have to solve and then later split it up.